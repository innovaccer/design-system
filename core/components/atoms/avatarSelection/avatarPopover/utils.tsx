import * as React from 'react';

export const handleKeyDown = (
  event: React.KeyboardEvent,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  setHighlightFirstItem?: React.Dispatch<React.SetStateAction<boolean>>,
  setHighlightLastItem?: React.Dispatch<React.SetStateAction<boolean>>,
  listRef?: any,
  withSearch?: boolean,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean>>,
  triggerRef?: any
) => {
  switch (event.key) {
    case 'ArrowUp':
      event.preventDefault();
      navigateOptions('up', focusedOption, setFocusedOption, listRef, withSearch);
      break;
    case 'ArrowDown':
      event.preventDefault();
      navigateOptions('down', focusedOption, setFocusedOption, listRef, withSearch);
      break;
    case 'Enter':
      handleEnterKey(focusedOption);
      setHighlightLastItem?.(false);
      setHighlightFirstItem?.(false);
      break;
    case 'Escape':
      setOpenPopover?.(false);
      triggerRef.current.focus();
      setFocusedOption?.(undefined);
      break;
    default:
      break;
  }
};

const handleEnterKey = (focusedOption: Element | undefined) => {
  (focusedOption as HTMLElement)?.click();
};

const navigateOptions = (
  direction: string,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  listRef?: any,
  withSearch?: boolean
) => {
  const listItems = listRef?.current.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let index = Array.from(listItems).findIndex((item) => {
    return item == focusedOption;
  });

  if (index === -1) {
    index = direction === 'up' ? listItems.length - 1 : 0;
  } else if (
    (withSearch && index === 0 && direction === 'up') ||
    (withSearch && index === listItems.length - 1 && direction === 'down')
  ) {
    const searchInput = listRef.current.querySelector('[data-test="DesignSystem-AvatarSelection--Input"]');
    searchInput.focus();
    setFocusedOption && setFocusedOption(searchInput);
  } else {
    index = direction === 'up' ? (index - 1 + listItems.length) % listItems.length : (index + 1) % listItems.length;

    const targetOption = listItems[index];

    (targetOption as HTMLElement).focus();
    setFocusedOption && setFocusedOption(targetOption);
    targetOption.scrollIntoView({ block: 'center' });
  }
};

export const handleInputKeyDown = (
  event: React.KeyboardEvent,
  listRef: any,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean>>,
  triggerRef?: any
) => {
  const listItems = listRef.current?.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let targetOption;

  switch (event.key) {
    case 'ArrowUp':
      event.preventDefault();
      targetOption = listItems[listItems.length - 1];
      break;
    case 'ArrowDown':
      event.preventDefault();
      targetOption = listItems[0];
      break;
    case 'Escape':
      setOpenPopover?.(false);
      triggerRef.current.focus();
      setFocusedOption?.(undefined);
      break;
    default:
      break;
  }

  (targetOption as HTMLElement)?.focus();
  targetOption?.scrollIntoView?.({ block: 'center' });
  setFocusedOption && setFocusedOption(targetOption);
};
