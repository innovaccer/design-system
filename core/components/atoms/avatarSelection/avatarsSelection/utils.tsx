import * as React from 'react';

export const handleKeyDown = (
  event: React.KeyboardEvent,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean>>,
  setHighlightFirstItem?: React.Dispatch<React.SetStateAction<boolean>>,
  setHighlightLastItem?: React.Dispatch<React.SetStateAction<boolean>>
) => {
  switch (event.key) {
    case 'Enter':
      setOpenPopover?.(true);
      setHighlightFirstItem?.(true);
      break;
    case 'ArrowDown':
      setOpenPopover?.(true);
      setHighlightFirstItem?.(true);
      break;
    case 'ArrowUp':
      setOpenPopover?.(true);
      setHighlightLastItem?.(true);
      break;
    default:
      break;
  }
};

export const focusListItem = (
  position: string,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  listRef?: any,
  withSearch?: boolean
) => {
  if (withSearch && position === 'down') {
    const searchInput = listRef.current?.querySelectorAll('[data-test="DesignSystem-AvatarSelection--Input"]');
    searchInput && searchInput[0]?.focus();
    setFocusedOption?.(searchInput);
    return;
  }

  const listItems = listRef.current?.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let targetOption;

  if (position === 'down') {
    targetOption = listItems?.[0];
  } else {
    targetOption = listItems?.[listItems.length - 1];
  }
  (targetOption as HTMLElement)?.focus();
  targetOption?.scrollIntoView?.({ block: 'center' });
  setFocusedOption && setFocusedOption(targetOption);
};
