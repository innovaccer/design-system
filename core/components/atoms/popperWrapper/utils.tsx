// Helper function to convert CSS class names to style objects
export const createStyleFromClass = (className: string) => {
  if (!className) return {};

  // Create a temporary element to get computed styles
  const tempElement = document.createElement('div');
  tempElement.className = className;
  document.body.appendChild(tempElement);

  const computedStyle = window.getComputedStyle(tempElement);
  const style: React.CSSProperties = {};

  // Extract relevant animation properties
  const animation = computedStyle.animation;
  const animationDuration = computedStyle.animationDuration;
  const animationTimingFunction = computedStyle.animationTimingFunction;
  const animationFillMode = computedStyle.animationFillMode;
  const opacity = computedStyle.opacity;
  const transform = computedStyle.transform;

  if (animation && animation !== 'none') {
    style.animation = animation;
  }
  if (animationDuration && animationDuration !== '0s') {
    style.animationDuration = animationDuration;
  }
  if (animationTimingFunction && animationTimingFunction !== 'ease') {
    style.animationTimingFunction = animationTimingFunction;
  }
  if (animationFillMode && animationFillMode !== 'none') {
    style.animationFillMode = animationFillMode;
  }
  if (opacity && opacity !== '1') {
    style.opacity = opacity;
  }
  if (transform && transform !== 'none') {
    style.transform = transform;
  }

  document.body.removeChild(tempElement);
  return style;
};
