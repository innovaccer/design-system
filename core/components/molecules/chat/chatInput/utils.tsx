import { MessageType } from './ChatInput';

export const positionMentionPopup = (
  textareaRef: React.RefObject<HTMLDivElement>,
  setMentionPosition: React.Dispatch<React.SetStateAction<{ top: number; left: number }>>
) => {
  if (!textareaRef.current) return;

  const selection = window.getSelection();
  if (!selection || !selection.rangeCount) return;

  const range = selection.getRangeAt(0);
  const rect = range.getBoundingClientRect();
  const editorRect = textareaRef.current.getBoundingClientRect();

  // Calculate position based on the cursor position
  let top = rect.bottom + window.scrollY;
  let left = rect.left + window.scrollX;

  // Adjust position if it would go off-screen
  if (top + 150 > window.innerHeight) {
    top = rect.top + window.scrollY - 150;
  }
  if (left + 200 > window.innerWidth) {
    left = window.innerWidth - 200;
  }

  // Adjust position relative to the editor
  top -= editorRect.top + window.scrollY - 12;
  left -= editorRect.left + window.scrollX - 8;

  setMentionPosition({ top, left });
};

export const extractMessageData = (textareaRef: React.RefObject<HTMLDivElement>): MessageType[] => {
  const elements = textareaRef.current?.childNodes;

  return Array.from(elements || []).map((item: any) => {
    if (item?.dataset?.type === 'mention') {
      const mentionValue = JSON.parse(item.dataset.content);
      return { type: 'mention', content: mentionValue?.data, nodeName: item.nodeName };
    }

    // handles new line by pressing enter
    else if (item.innerText === '\n' && item.nodeName === 'DIV') {
      return { type: 'newline', content: '\n', nodeName: item.nodeName };
    }

    // handles line break by pressing shift + enter
    else if (item.nodeName === 'BR') {
      return { type: 'linebreak', content: '<br>', nodeName: item.nodeName };
    }

    return { type: 'text', content: item.innerText || item.textContent, nodeName: item.nodeName };
  });
};
