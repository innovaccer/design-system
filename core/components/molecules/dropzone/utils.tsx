const isIe = (userAgent: string) => {
  return userAgent.indexOf('MSIE') !== -1 || userAgent.indexOf('Trident/') !== -1;
};

const isEdge = (userAgent: string) => {
  return userAgent.indexOf('Edge/') !== -1;
};

export const accepts = (file: File, acceptedFiles?: string | string[]) => {
  if (file && acceptedFiles) {
    const acceptedFilesArray = Array.isArray(acceptedFiles) ? acceptedFiles : acceptedFiles.split(',');
    const fileName = file.name || '';
    const mimeType = (file.type || '').toLowerCase();
    const baseMimeType = mimeType.replace(/\/.*$/, '');

    if (!mimeType) {
      return true;
    }

    return acceptedFilesArray.some((type) => {
      const validType = type.trim().toLowerCase();
      if (validType.charAt(0) === '.') {
        return fileName.toLowerCase().endsWith(validType);
      }
      if (validType.endsWith('/*')) {
        return baseMimeType === validType.replace(/\/.*$/, '');
      }
      return mimeType === validType;
    });
  }
  return true;
};

export const isPropagationStopped = (event: any) => {
  if (typeof event.isPropagationStopped === 'function') {
    return event.isPropagationStopped();
  }
  if (typeof event.cancelBubble !== 'undefined') {
    return event.cancelBubble;
  }
  return false;
};

export const isEvtWithFiles = (event: any) => {
  if (!event.dataTransfer) {
    return !!event.target && !!event.target.files;
  }

  return Array.prototype.some.call(
    event.dataTransfer.types,
    (type) => type === 'Files' || type === 'application/x-moz-file'
  );
};

export const isKindFile = (item: null | Record<string, unknown>) => {
  return typeof item === 'object' && item !== null && item.kind === 'file';
};

export const onDocumentDragOver = (event: DragEvent) => {
  event.preventDefault();
};

export const isIeOrEdge = (userAgent = window.navigator.userAgent) => {
  return isIe(userAgent) || isEdge(userAgent);
};

type functype = (event: Event, ...args: unknown[]) => void;

export const composeEventHandlers = (...fns: functype[]) => {
  return (event: Event, ...args: unknown[]) =>
    fns.some((fn) => {
      if (!isPropagationStopped(event) && fn) {
        fn(event, ...args);
      }
      return isPropagationStopped(event);
    });
};

export const reducer = (state: any, action: any) => {
  switch (action.type) {
    case 'focus':
      return {
        ...state,
        isFocused: true,
      };
    case 'blur':
      return {
        ...state,
        isFocused: false,
      };
    case 'openDialog':
      return {
        ...state,
        isFileDialogActive: true,
      };
    case 'closeDialog':
      return {
        ...state,
        isFileDialogActive: false,
      };
    case 'setDraggedFiles':
      /* eslint no-case-declarations: 0 */
      const { isDragActive, draggedFiles } = action;
      return {
        ...state,
        draggedFiles,
        isDragActive,
      };
    case 'setFiles':
      return {
        ...state,
        acceptedFiles: action.acceptedFiles,
        fileRejections: action.fileRejections,
      };
    case 'reset':
      return {
        ...state,
        isFileDialogActive: false,
        isDragActive: false,
        draggedFiles: [],
        acceptedFiles: [],
        fileRejections: [],
      };
    default:
      return state;
  }
};
