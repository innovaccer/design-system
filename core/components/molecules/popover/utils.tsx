import { Placement } from '@floating-ui/react';
import { PositionType as Position } from '@/common.type';

export const mapPositionToPlacement = (position: Position): Placement => {
  const positionMap: Record<Position, Placement> = {
    auto: 'bottom',
    'auto-start': 'bottom-start',
    'auto-end': 'bottom-end',
    'top-start': 'top-start',
    top: 'top',
    'top-end': 'top-end',
    'right-start': 'right-start',
    right: 'right',
    'right-end': 'right-end',
    'bottom-end': 'bottom-end',
    bottom: 'bottom',
    'bottom-start': 'bottom-start',
    'left-end': 'left-end',
    left: 'left',
    'left-start': 'left-start',
  };

  return positionMap[position] || 'bottom';
};
