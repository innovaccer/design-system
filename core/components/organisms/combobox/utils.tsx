import * as React from 'react';

export const handleKeyDown = (
  event: React.KeyboardEvent,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean>>,
  inputTriggerRef?: any,
  setHighlightFirstItem?: React.Dispatch<React.SetStateAction<boolean>>,
  setHighlightLastItem?: React.Dispatch<React.SetStateAction<boolean>>,
  multiSelect?: boolean,
  listRef?: any
) => {
  switch (event.key) {
    case 'ArrowUp':
      event.preventDefault();
      navigateOptions('up', focusedOption, setFocusedOption, listRef);
      break;
    case 'ArrowDown':
      event.preventDefault();
      navigateOptions('down', focusedOption, setFocusedOption, listRef);
      break;
    case 'Enter':
      handleEnterKey(focusedOption, multiSelect, inputTriggerRef, listRef, setFocusedOption);
      setHighlightLastItem?.(false);
      setHighlightFirstItem?.(false);
      break;
    case 'Escape':
      setOpenPopover?.(false);
      inputTriggerRef.current.focus();
      setFocusedOption?.(undefined);
      break;
    default:
      break;
  }
};

const handleEnterKey = (
  focusedOption: Element | undefined,
  multiSelect?: boolean,
  inputTriggerRef?: any,
  listRef?: any,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>
) => {
  (focusedOption as HTMLElement)?.click();
  if (!multiSelect) {
    inputTriggerRef.current.focus();
  } else {
    // to focus first option by default when last option is selected
    const listItems = listRef.current.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
    const listArr = Array.from(listItems);
    const index = listArr.findIndex((item) => {
      return item == focusedOption;
    });

    if (index === listArr.length - 1) {
      (listItems[0] as HTMLElement).focus();
      setFocusedOption && setFocusedOption(listItems[0]);
      listItems[0].scrollIntoView({ block: 'center' });
    }
  }
};

const navigateOptions = (
  direction: string,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  listRef?: any
) => {
  const listItems = listRef.current.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let index = Array.from(listItems).findIndex((item) => {
    return item == focusedOption;
  });

  if (index === -1) {
    index = direction === 'up' ? listItems.length - 1 : 0;
  } else {
    index = direction === 'up' ? (index - 1 + listItems.length) % listItems.length : (index + 1) % listItems.length;
  }

  const targetOption = listItems[index];

  (targetOption as HTMLElement).focus();
  setFocusedOption && setFocusedOption(targetOption);
  targetOption?.scrollIntoView?.({ block: 'center' });
};
