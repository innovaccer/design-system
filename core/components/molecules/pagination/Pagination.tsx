import * as React from 'react';
import { Text, MetricInput, Button } from '@/index';
import { BaseProps, extractBaseProps } from '@/utils/types';

import classNames from 'classnames';

export type PaginationType = 'basic' | 'jump';

export interface PaginationProps extends BaseProps {
  /**
   * `Pagination` component type
   */
  type: PaginationType;
  /**
   * Total number of pages
   */
  totalPages: number;
  /**
   * Current page
   */
  page: number;
  /**
   *  Callback when page is changed
   */
  onPageChange: (page: number) => void;
}

export const Pagination = (props: PaginationProps) => {
  const { type, totalPages, onPageChange, className } = props;

  const baseProps = extractBaseProps(props);

  const [page, setPage] = React.useState<number>(props.page);
  const [init, setInit] = React.useState<boolean>(false);

  React.useEffect(() => {
    setPage(props.page);
  }, [props.page]);

  const wrapperClass = classNames(
    {
      ['Pagination']: true,
      [`Pagination--${type}`]: type,
    },
    className
  );

  const nextButtonWrapperClass = classNames({
    ['Pagination-buttonWrapper']: true,
    ['Pagination-buttonWrapper--next']: true,
  });

  const prevButtonWrapperClass = classNames({
    ['Pagination-buttonWrapper']: true,
    ['Pagination-buttonWrapper--previous']: true,
  });

  React.useEffect(() => {
    if (init) {
      if (page >= 1 && page <= totalPages) onPageChange(page);
    }
  }, [page]);

  const inputChangeHandler = (e: React.ChangeEvent<HTMLInputElement>) => {
    e.preventDefault();

    const val = +e.target.value.trim();
    if (val >= 0 && val <= totalPages) {
      if (!init) setInit(true);
      setPage(val);
    }
  };

  const onClickHandler = (buttonType: 'prev' | 'next' | 'first' | 'last') => {
    setInit(true);
    switch (buttonType) {
      case 'first':
        setPage(1);
        break;
      case 'last':
        setPage(totalPages);
        break;
      case 'prev':
        if (page > 1) setPage(page - 1);
        break;
      case 'next':
        if (page < totalPages) setPage(page + 1);
        break;
    }
  };

  return (
    <div data-test="DesignSystem-Pagination" {...baseProps} className={wrapperClass}>
      <div className={prevButtonWrapperClass}>
        <Button
          onClick={() => onClickHandler('first')}
          disabled={page <= 1}
          appearance="transparent"
          icon="first_page"
          data-test="DesignSystem-Pagination--FirstButton"
        />
        <Button
          onClick={() => onClickHandler('prev')}
          disabled={page <= 1}
          icon="navigate_before"
          data-test="DesignSystem-Pagination--PrevButton"
          className="ml-4 mr-3"
        />
      </div>
      {type === 'jump' && (
        <div className="Pagination-pageIndex">
          <MetricInput
            name="page"
            className="Pagination-MetricInput"
            onChange={inputChangeHandler}
            value={`${page === 0 ? '' : page}`}
            data-test="DesignSystem-Pagination--Input"
          />
          <Text>{` of ${totalPages} pages`}</Text>
        </div>
      )}
      <div className={nextButtonWrapperClass}>
        <Button
          onClick={() => onClickHandler('next')}
          disabled={page >= totalPages}
          icon="navigate_next"
          data-test="DesignSystem-Pagination--NextButton"
          className="mr-4 ml-3"
        />
        <Button
          onClick={() => onClickHandler('last')}
          disabled={page >= totalPages}
          appearance="transparent"
          icon="last_page"
          data-test="DesignSystem-Pagination--LastButton"
        />
      </div>
    </div>
  );
};

Pagination.displayName = 'Pagination';
Pagination.defaultProps = {
  type: 'basic',
  page: 1,
  totalPages: 1,
};

export default Pagination;
