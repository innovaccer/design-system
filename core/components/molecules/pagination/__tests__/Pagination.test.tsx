import * as React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Pagination, { PaginationProps as Props } from '../Pagination';
import { testHelper, filterUndefined, valueHelper, testMessageHelper } from '@/utils/testHelper';

const pageNumber = 10;
const totalPages = 50;
const paginationType = ['basic', 'jump'];
const FunctionValue = jest.fn();

const inputTestId = 'DesignSystem-Pagination--Input';

describe('Pagination component', () => {
  const mapper: Record<string, any> = {
    type: valueHelper(paginationType, { required: true, iterate: true }),
    page: valueHelper(pageNumber, { required: true }),
    totalPages: valueHelper(totalPages, { required: true }),
    onPageChange: valueHelper(FunctionValue, { required: true }),
  };

  const testFunc = (props: Record<string, any>): void => {
    const attr = filterUndefined(props) as Props;

    it(testMessageHelper(attr), () => {
      const { asFragment } = render(<Pagination {...attr} />);
      expect(asFragment()).toMatchSnapshot();
    });
  };

  testHelper(mapper, testFunc);
});

describe('Pagination component with props: page and onPageChange', () => {
  it('should render currentPage initially', () => {
    const { getByTestId } = render(
      <Pagination totalPages={totalPages} onPageChange={FunctionValue} type="jump" page={pageNumber} />
    );

    expect(getByTestId(inputTestId)).toHaveValue(pageNumber);
  });

  it('should render correct total page count', () => {
    const { getByTestId } = render(
      <Pagination page={pageNumber} totalPages={totalPages} type="jump" onPageChange={FunctionValue} />
    );

    expect(getByTestId('DesignSystem-Pagination').textContent).toMatch(`${totalPages}`);
  });

  it('render page due to Input change', () => {
    const updatedPage = 15;
    const { getByTestId } = render(
      <Pagination page={pageNumber} totalPages={totalPages} type="jump" onPageChange={FunctionValue} />
    );

    const input = getByTestId(inputTestId);

    fireEvent.change(input, { target: { value: updatedPage } });
    expect(getByTestId(inputTestId)).toHaveValue(updatedPage);
    expect(FunctionValue).toHaveBeenCalledWith(updatedPage);

    fireEvent.change(input, { target: { value: 0 } });
    expect(getByTestId(inputTestId)).not.toHaveValue(0);

    fireEvent.change(input, { target: { value: totalPages + 1 } });
    expect(getByTestId(inputTestId)).not.toHaveValue(totalPages + 1);
  });

  it('current page should be incremented on click of next button', () => {
    const { getByTestId } = render(
      <Pagination page={pageNumber} totalPages={totalPages} type="jump" onPageChange={FunctionValue} />
    );

    const nextButton = getByTestId('DesignSystem-Pagination--NextButton');
    fireEvent.click(nextButton);

    expect(getByTestId(inputTestId)).toHaveValue(pageNumber + 1);
    expect(FunctionValue).toHaveBeenCalledWith(pageNumber + 1);
  });

  it('current page should be decremented on click of previous button', () => {
    const { getByTestId } = render(
      <Pagination page={pageNumber} totalPages={totalPages} type="jump" onPageChange={FunctionValue} />
    );

    const prevButton = getByTestId('DesignSystem-Pagination--PrevButton');
    fireEvent.click(prevButton);

    expect(getByTestId(inputTestId)).toHaveValue(pageNumber - 1);
    expect(FunctionValue).toHaveBeenCalledWith(pageNumber - 1);
  });

  it('current page should be last page number when clicked on last page icon', () => {
    const { getByTestId } = render(
      <Pagination page={pageNumber} totalPages={totalPages} type="jump" onPageChange={FunctionValue} />
    );

    const lastButton = getByTestId('DesignSystem-Pagination--LastButton');
    fireEvent.click(lastButton);

    expect(getByTestId(inputTestId)).toHaveValue(totalPages);
    expect(FunctionValue).toHaveBeenCalledWith(totalPages);
  });

  it('current page should be first page when clicked on first page icon', () => {
    const { getByTestId } = render(
      <Pagination page={pageNumber} totalPages={totalPages} type="jump" onPageChange={FunctionValue} />
    );

    const firstButton = getByTestId('DesignSystem-Pagination--FirstButton');
    fireEvent.click(firstButton);

    expect(getByTestId(inputTestId)).toHaveValue(1);
    expect(FunctionValue).toHaveBeenCalledWith(1);
  });

  it('current page should be first page if page no. is 1 and previous button is clicked', () => {
    const pageValue = 1;

    const { getByTestId } = render(
      <Pagination page={pageValue} totalPages={totalPages} type="jump" onPageChange={FunctionValue} />
    );

    const prevButton = getByTestId('DesignSystem-Pagination--PrevButton');
    fireEvent.click(prevButton);

    expect(getByTestId(inputTestId)).toHaveValue(pageValue);
    expect(getByTestId('DesignSystem-Pagination--FirstButton')).toHaveAttribute('disabled');
    expect(getByTestId('DesignSystem-Pagination--PrevButton')).toHaveAttribute('disabled');
  });

  it('current page should be last page even if page >= totalPage', () => {
    const pageValue = totalPages;

    const { getByTestId } = render(
      <Pagination page={pageValue} totalPages={totalPages} type="jump" onPageChange={FunctionValue} />
    );

    const nextButton = getByTestId('DesignSystem-Pagination--NextButton');
    fireEvent.click(nextButton);

    expect(getByTestId(inputTestId)).toHaveValue(pageValue);
    expect(getByTestId('DesignSystem-Pagination--LastButton')).toHaveAttribute('disabled');
    expect(getByTestId('DesignSystem-Pagination--NextButton')).toHaveAttribute('disabled');
  });

  it('should update current page on change of prop: page', () => {
    const updatedPage = 20;

    const { getByTestId, rerender } = render(
      <Pagination page={pageNumber} totalPages={totalPages} type="jump" onPageChange={FunctionValue} />
    );

    expect(getByTestId(inputTestId)).toHaveValue(pageNumber);

    rerender(<Pagination page={updatedPage} totalPages={totalPages} type="jump" onPageChange={FunctionValue} />);

    expect(getByTestId(inputTestId)).toHaveValue(updatedPage);
  });
});

describe('Pagination component with prop: type', () => {
  it('renders default prop: type', () => {
    const { getByTestId } = render(<Pagination totalPages={totalPages} onPageChange={FunctionValue} />);

    expect(getByTestId('DesignSystem-Pagination')).toHaveClass('Pagination--basic');
    expect(getByTestId('DesignSystem-Pagination--PrevButton')).toHaveClass('ml-4 mr-3');
    expect(getByTestId('DesignSystem-Pagination').textContent).not.toMatch(`${totalPages}`);
  });

  it('renders default prop: type', () => {
    const { getByTestId } = render(<Pagination totalPages={totalPages} onPageChange={FunctionValue} type="jump" />);

    expect(getByTestId('DesignSystem-Pagination')).toHaveClass('Pagination--jump');
    expect(getByTestId('DesignSystem-Pagination--NextButton')).toHaveClass('mr-4 ml-3');
  });
});

describe('Pagination Component with overwrite class', () => {
  const className = 'DS-Pagination';

  it('overwrite Pagination class', () => {
    const { getByTestId } = render(
      <Pagination className={className} totalPages={totalPages} onPageChange={FunctionValue} />
    );
    expect(getByTestId('DesignSystem-Pagination')).toHaveClass(className);
  });
});
