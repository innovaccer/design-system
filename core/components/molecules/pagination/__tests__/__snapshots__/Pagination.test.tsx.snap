// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination component 
  type: "basic", page: 1, totalPages: 50, onPageChange: "[Function]"
 1`] = `
<div
  className="Pagination Pagination--basic"
>
  <Button
    appearance="transparent"
    disabled={true}
    icon="first_page"
    onClick={[Function]}
    size="large"
  />
  <div
    className="ml-4 mx-3"
  >
    <Button
      disabled={true}
      icon="navigate_before"
      onClick={[Function]}
      size="large"
    />
  </div>
  <div
    className="mr-4 mx-3"
  >
    <Button
      disabled={false}
      icon="navigate_next"
      onClick={[Function]}
      size="large"
    />
  </div>
  <Button
    appearance="transparent"
    disabled={false}
    icon="last_page"
    onClick={[Function]}
    size="large"
  />
</div>
`;

exports[`Pagination component 
  type: "basic", page: 5, totalPages: 50, onPageChange: "[Function]"
 1`] = `
<div
  className="Pagination Pagination--basic"
>
  <Button
    appearance="transparent"
    disabled={false}
    icon="first_page"
    onClick={[Function]}
    size="large"
  />
  <div
    className="ml-4 mx-3"
  >
    <Button
      disabled={false}
      icon="navigate_before"
      onClick={[Function]}
      size="large"
    />
  </div>
  <div
    className="mr-4 mx-3"
  >
    <Button
      disabled={false}
      icon="navigate_next"
      onClick={[Function]}
      size="large"
    />
  </div>
  <Button
    appearance="transparent"
    disabled={false}
    icon="last_page"
    onClick={[Function]}
    size="large"
  />
</div>
`;

exports[`Pagination component 
  type: "basic", page: 10, totalPages: 50, onPageChange: "[Function]"
 1`] = `
<div
  className="Pagination Pagination--basic"
>
  <Button
    appearance="transparent"
    disabled={false}
    icon="first_page"
    onClick={[Function]}
    size="large"
  />
  <div
    className="ml-4 mx-3"
  >
    <Button
      disabled={false}
      icon="navigate_before"
      onClick={[Function]}
      size="large"
    />
  </div>
  <div
    className="mr-4 mx-3"
  >
    <Button
      disabled={true}
      icon="navigate_next"
      onClick={[Function]}
      size="large"
    />
  </div>
  <Button
    appearance="transparent"
    disabled={true}
    icon="last_page"
    onClick={[Function]}
    size="large"
  />
</div>
`;

exports[`Pagination component 
  type: "basic", page: 15, totalPages: 50, onPageChange: "[Function]"
 1`] = `
<div
  className="Pagination Pagination--basic"
>
  <Button
    appearance="transparent"
    disabled={false}
    icon="first_page"
    onClick={[Function]}
    size="large"
  />
  <div
    className="ml-4 mx-3"
  >
    <Button
      disabled={false}
      icon="navigate_before"
      onClick={[Function]}
      size="large"
    />
  </div>
  <div
    className="mr-4 mx-3"
  >
    <Button
      disabled={false}
      icon="navigate_next"
      onClick={[Function]}
      size="large"
    />
  </div>
  <Button
    appearance="transparent"
    disabled={false}
    icon="last_page"
    onClick={[Function]}
    size="large"
  />
</div>
`;

exports[`Pagination component 
  type: "jump", page: 1, totalPages: 50, onPageChange: "[Function]"
 1`] = `
<div
  className="Pagination Pagination--jump"
>
  <Button
    appearance="transparent"
    disabled={true}
    icon="first_page"
    onClick={[Function]}
    size="large"
  />
  <div
    className="ml-4 mx-4"
  >
    <Button
      disabled={true}
      icon="navigate_before"
      onClick={[Function]}
      size="large"
    />
  </div>
  <div
    className="Pagination-pageIndex"
  >
    <Input
      clearButton={false}
      name="page"
      onChange={[Function]}
      size="large"
      type="number"
      value="1"
    />
    <Text>
       of 10 pages
    </Text>
  </div>
  <div
    className="mr-4 mx-4"
  >
    <Button
      disabled={false}
      icon="navigate_next"
      onClick={[Function]}
      size="large"
    />
  </div>
  <Button
    appearance="transparent"
    disabled={false}
    icon="last_page"
    onClick={[Function]}
    size="large"
  />
</div>
`;

exports[`Pagination component 
  type: "jump", page: 5, totalPages: 50, onPageChange: "[Function]"
 1`] = `
<div
  className="Pagination Pagination--jump"
>
  <Button
    appearance="transparent"
    disabled={false}
    icon="first_page"
    onClick={[Function]}
    size="large"
  />
  <div
    className="ml-4 mx-4"
  >
    <Button
      disabled={false}
      icon="navigate_before"
      onClick={[Function]}
      size="large"
    />
  </div>
  <div
    className="Pagination-pageIndex"
  >
    <Input
      clearButton={false}
      name="page"
      onChange={[Function]}
      size="large"
      type="number"
      value="5"
    />
    <Text>
       of 10 pages
    </Text>
  </div>
  <div
    className="mr-4 mx-4"
  >
    <Button
      disabled={false}
      icon="navigate_next"
      onClick={[Function]}
      size="large"
    />
  </div>
  <Button
    appearance="transparent"
    disabled={false}
    icon="last_page"
    onClick={[Function]}
    size="large"
  />
</div>
`;

exports[`Pagination component 
  type: "jump", page: 10, totalPages: 50, onPageChange: "[Function]"
 1`] = `
<div
  className="Pagination Pagination--jump"
>
  <Button
    appearance="transparent"
    disabled={false}
    icon="first_page"
    onClick={[Function]}
    size="large"
  />
  <div
    className="ml-4 mx-4"
  >
    <Button
      disabled={false}
      icon="navigate_before"
      onClick={[Function]}
      size="large"
    />
  </div>
  <div
    className="Pagination-pageIndex"
  >
    <Input
      clearButton={false}
      name="page"
      onChange={[Function]}
      size="large"
      type="number"
      value="10"
    />
    <Text>
       of 10 pages
    </Text>
  </div>
  <div
    className="mr-4 mx-4"
  >
    <Button
      disabled={true}
      icon="navigate_next"
      onClick={[Function]}
      size="large"
    />
  </div>
  <Button
    appearance="transparent"
    disabled={true}
    icon="last_page"
    onClick={[Function]}
    size="large"
  />
</div>
`;

exports[`Pagination component 
  type: "jump", page: 15, totalPages: 50, onPageChange: "[Function]"
 1`] = `
<div
  className="Pagination Pagination--jump"
>
  <Button
    appearance="transparent"
    disabled={false}
    icon="first_page"
    onClick={[Function]}
    size="large"
  />
  <div
    className="ml-4 mx-4"
  >
    <Button
      disabled={false}
      icon="navigate_before"
      onClick={[Function]}
      size="large"
    />
  </div>
  <div
    className="Pagination-pageIndex"
  >
    <Input
      clearButton={false}
      name="page"
      onChange={[Function]}
      size="large"
      type="number"
      value="15"
    />
    <Text>
       of 10 pages
    </Text>
  </div>
  <div
    className="mr-4 mx-4"
  >
    <Button
      disabled={false}
      icon="navigate_next"
      onClick={[Function]}
      size="large"
    />
  </div>
  <Button
    appearance="transparent"
    disabled={false}
    icon="last_page"
    onClick={[Function]}
    size="large"
  />
</div>
`;
